src/backend/cdb/cdbquerycontextdispatching.c

```diff
@@ -779,21 +779,24 @@ static void
 RebuildNamespace(QueryContextInfo *cxt)
 {
 
+	MemoryContext oldContext;
 	int len;
 	char buffer[4], *binary;
+	oldContext = MemoryContextSwitchTo(MessageContext);
 	ReadData(cxt, buffer, sizeof(buffer), TRUE);
 
 	len = (int) ntohl(*(uint32 *) buffer);
 	binary = palloc(len);
 	if(ReadData(cxt, binary, len, TRUE))
 	{
-		StringInfoData buffer;
-		initStringInfoOfString(&buffer, binary, len);
-		dfs_address = strdup(buffer.data);
+		dfs_address = pstrdup(binary);
```

Comment:
In the definition of `pstrdup` it states 

```
/*
 * MemoryContextStrdup
 *		Like strdup(), but allocate from the specified context
 */
```

I chose to use the intermediary `binary` so that if there is an error `dfs_address` is not set with any data, but your point for not using `binary` at all is valid.

We are aware that we should `pfree` even with MemoryContext, but we need the `dfs_address` for the lifetime of each query so this is the simplest & safest means of doing so with the time given